_python/main/views.py

```diff
@@ -256,75 +260,134 @@ def dashboard(request, user_id):
+            return HttpResponseRedirect(canonical)
 
-        Other users cannot see draft casebooks:
-        >>> check_response(client.get(draft_casebook.get_absolute_url()), status_code=302)
-        >>> check_response(client.get(draft_casebook.get_absolute_url(), as_user=non_collaborating_user), status_code=403)
-    """
-
-    casebook = get_object_or_404(Casebook, id=casebook_param['id'])
-
-    # check permissions
-    if not casebook.viewable_by(request.user):
-        return login_required_response(request)
+        contents = casebook.contents.prefetch_resources()
+        return render_with_actions(request, 'casebook.html', {
+            'casebook': casebook,
+            'contents': contents
+        })
 
-    # canonical redirect
-    canonical = casebook.get_absolute_url()
-    if request.path != canonical:
-        return HttpResponseRedirect(canonical)
+    @method_decorator(login_required)
+    def patch(self, request, casebook_param):
+        """
+            Publish a casebook.
+
+            Given:
+            >>> casebook, casebook_factory, client, admin_user, user_factory = [getfixture(f) for f in ['casebook', 'casebook_factory', 'client', 'admin_user', 'user_factory']]
+            >>> user = casebook.collaborators.first()
+            >>> non_collaborating_user = user_factory()
+            >>> private_casebook = casebook_factory(contentcollaborator_set__user=user, public=False)
+            >>> draft_casebook = casebook_factory(contentcollaborator_set__user=user, public=False, draft_mode_of_published_casebook=True, copy_of=casebook)
+
+            Only authors (admins, etc.) may publish casebooks.
+            >>> check_response(client.patch(draft_casebook.get_absolute_url()), status_code=302)
+            >>> check_response(client.patch(draft_casebook.get_absolute_url(), as_user=non_collaborating_user), status_code=403)
+
+            Already-published casebooks cannot be re-published.
+            >>> check_response(client.patch(casebook.get_absolute_url(), as_user=user), status_code=403)
+
+            Newly-composed (private, never-published) casebooks, when published, become public.
+            >>> check_response(
+            ...     client.patch(private_casebook.get_absolute_url(), as_user=user, follow=True),
+            ...     expected_url=private_casebook.get_absolute_url(),
+            ...     content_includes=private_casebook.title,
+            ...     content_excludes="You are viewing a preview"
+            ... )
+            >>> private_casebook.refresh_from_db()
+            >>> assert private_casebook.is_public
+
+            Drafts of already-published casebooks, when published, replace their parent.
+            # NB: calculating expected_url this way is a hack, since we don't know it
+            # when check_response is called.... can we think of a better strategy? Do we
+            # want `check_response` to return the response, for optional further testing?
```

Comment:
I briefly had `check_response` return response, but I ended up preferring this:

```
>>> response = client.patch(draft_casebook.get_absolute_url(), as_user=user, follow=True)
>>> check_response(response, ...)
>>> ...
>>> assertURLEqual(response, casebook.get_absolute_url())
```

... so you don't need to remember `assert check_response()` in the typical case. This often ends up as short or shorter as check_response needs less whitespace. I'd be fine with requiring `assert check_response()` though if you like that approach.

Reply:
Ah! Totally. For whatever reason, it just did not occur to me to call `client` separately first. I'll revise this to do that.